server/src/main/java/org/elasticsearch/ingest/IngestDocMetadata.java

```diff
@@ -100,4 +117,16 @@ private static void versionTypeValidator(String key, String value) {
                 + "]"
         );
     }
+
+    @Override
+    public boolean isAvailable(String key) {
+        // the key cannot be null or empty because of the nature of the calling code, and this is already validated in IngestDocument
+        assert key != null && key.isEmpty() == false;
+        // we can avoid a map lookup on most keys since we know that the only keys that are 'metadata keys' for an ingest document
+        // must be keys that start with an underscore
+        if (key.charAt(0) != UNDERSCORE) {
```

Comment:
I'm amazed that this makes a difference at all (and I'm really curious why), but I've seen the charts!